File: javascript/permute.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let visited = new Set();
    let perms = [];
    let helper = (perm = []) => {
        for (let i = 0; i < nums.length; i++) {
            if (!visited.has(nums[i])) {
                visited.add(nums[i]);
                perm.push(nums[i]);
                if (perm.length === nums.length) {
                    perms.push(perm.slice());
                } else {
                    helper(perm);
                }
                visited.delete(nums[i]);
                perm.pop();
            }
        }
    }
    helper();
    return perms;
};

File: python/permute.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        visited = set()
        perms = []
        def helper(perm):
            for x in range(0, len(nums)):
                if nums[x] not in visited:
                    visited.add(nums[x])
                    perm.append(nums[x])
                    if len(perm) == len(nums):
                        perms.append(perm[0:])
                    else:
                        helper(perm)
                    visited.remove(nums[x])
                    perm.pop()
        helper([])
        return perms
        
